VectorInstitute/FL4Health
# Federated ensemble

# PR Type
[**Feature** | Fix | Documentation | Other ]

# Short Description

Clickup Ticket(s): [Federated training ensemble of models](https://app.clickup.com/t/8686ja6he)

Small PR that adds some initial functionality to support training ensembles of models in a federated manner. To this end, I added two classes: 
- `EnsembleModel`: A nn.Module that acts as a wrapper for the ensemble of models with support to produce predictions for individual models as well as an ensemble prediction via voting or averaging.
- `EnsembleClient`: A custom client to handle specific details of training a federated ensemble of models such as setting up the client, computing losses and performing a train step. 

In the process I had to make some minor changes to the `BasicClient` such as store client optimizers in a dictionary  and making the backward attribute of `Losses` a union of Tensor and a dictionary of Tensors. 

It is worth mentioning that I plan on doing a follow up PR that makes it so that we can use Ensembling in conjunction with other client types (ie Scaffold). This seems like a good first start and it would be good to get some eyes on it before I go further. 

# Tests Added
- Add tests to ensure the `EnsembleModel` produces predictions of the current dimension
- Add tests to ensure the "ensemble prediction methods" (ie voting and averaging) work correctly
- A small addition to the tests on `Losses` focused on the new functionality I mentioned
